src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ObjectNode.Sorting.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+using ILCompiler.DependencyAnalysisFramework;
+using Internal.TypeSystem;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    partial class ObjectNode
+    {
+        protected abstract int ClassCode { get; }
+
+        protected abstract int CompareToImpl(ObjectNode other, TypeSystemComparer comparer, NameMangler nameMangler);
+
+        protected enum ObjectNodeOrder
+        {
+            //
+            // The ordering of this sequence of nodes is deliberate and currently required for 
+            // compiler correctness.
+            //
+            MetadataNode,
+            ResourceDataNode,
+            ResourceIndexNode,
+            TypeMetadataMapNode,
+            ClassConstructorContextMap,
+            DynamicInvokeTemplateDataNode,
+            ReflectionInvokeMapNode,
+            DelegateMarshallingStubMapNode,
+            StructMarshallingStubMapNode,
+            ArrayMapNode,
+            ReflectionFieldMapNode,
+            NativeLayoutInfoNode,
+            ExactMethodInstantiationsNode,
+            GenericTypesHashtableNode,
+            GenericMethodsHashtableNode,
+            GenericVirtualMethodTableNode,
+            InterfaceGenericVirtualMethodTableNode,
+            GenericMethodsTemplateMap,
+            GenericTypesTemplateMap,
+            BlockReflectionTypeMapNode,
+            StaticsInfoHashtableNode,
+            ReflectionVirtualInvokeMapNode,
+            ExternalReferencesTableNode,
+
+            //
+            // The ordering of this sequence of nodes is arbitrary.
```

Comment:
If you have a plan for supporting node types that only exist on the TFS side without necessarily putting them in this file, I'm fine with this.